projects/transformers/run.py

```diff
@@ -233,20 +222,104 @@ def run_pretraining(
         tokenizer=tokenizer, mlm_probability=data_args.mlm_probability
     )
 
-    # Train and evaluate
-    trainer = init_trainer(
-        model, tokenizer, data_collator, training_args,
-        train_dataset, eval_dataset,
-        trainer_class=model_args.trainer_class,
-        trainer_extra_kwargs=model_args.trainer_extra_kwargs,
-        trainer_callbacks=model_args.trainer_callbacks or None,
-    )
-    if training_args.do_train:
-        train(trainer, training_args.output_dir, last_checkpoint)
+    # Initialize distillation models
+    if model_args.teacher_models_name_or_path:
+        teacher_models = []
+        for model_name_or_path in model_args.teacher_models_name_or_path:
+            teacher_args = replace(model_args, model_name_or_path=model_name_or_path)
+            teacher_config = init_config(teacher_args)
+            teacher_models.append(init_model(teacher_args, teacher_config, tokenizer))
+
+        model_args.trainer_extra_kwargs["teacher_models"] = teacher_models
+        logging.info(f"{len(teacher_models)} teacher models initialized.")
+
+    # Regular training run: init model, trainer, and call train
+    if model_args.hp_num_trials <= 1:
```

Comment:
I see your point on the readability. I can move the code inside the if into  a `train_with_hpsearch` in run_utils. I won't refactor the whole function to have two differents runs, though, that is not the purpose of this PR and I don't think it will be cleaner or useful to have that amount of code repetition.

Reply:
Done